rfc/0020-group-policies.md

```diff
@@ -0,0 +1,263 @@
+
+This avoids evaluating all the policies in the group when the result is already known.
+
+### Settings validation
+
+When the policy server starts, it will validate the settings of the group policies as well as the single policies.
+However, the group policies will have an additional validation step to ensure that the expression is valid and
+that evaluates to a boolean value.
+
+### Handler
+
+The handler will respond to the `/validate/<group name>` endpoint.
+There is no need to create a new endpoint for the group policies, and as far as the API is concerned, the group policies will be treated as single policies.
+
+After the validation step, precompiled policies and group settings will be added to the `EvaluationEnvironment`.
+When a validation request is received, if the policy is a group policy, the `EvaluationEnvironment` performs the following steps:
+
+- create a new Rhai [raw Engine](https://rhai.rs/book/engine/raw.html?highlight=raw%20eng#raw-engine)
+- bind native functions to the engine that validate the policy
+- call the engine eval method with the expression
+- return the result
+
+Creating a new engine for each request is desirable because it avoids contention and locking of the Engine instance.
+This pattern is described in the Rhai book, under the [One Engine Instance Per Call](https://rhai.rs/book/patterns/parallel.html?highlight=one%20engine%20per#one-engine-instance-per-call) section.
+
+To improve performance, we could consider precompiling the AST of the expression when building the `EvaluationEnvironment`.
+Also, we could consider creating a [custom package](https://rhai.rs/book/rust/packages/create.html?highlight=custom%20package#create-a-custom-package) to instantiate the policy function calls only once.
+
+### Preliminary benchmark
+
+The following results are based on the [Kubewarden k6 load test](https://github.com/kubewarden/load-testing/tree/k6) using the [psp-apparmor policy](https://github.com/kubewarden/apparmor-psp-policy)
+
+|               |                          | http_req_duration |
+| ------------- | ------------------------ | ----------------- |
+| single policy | psp-apparmor             | avg=2.02ms        |
+| group policy  | `psp-apparmor() && true` | avg=2.46ms        |
+
+Note that no optimizations were made to the Rhai engine in this POC implementation.
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+- The implementation of the group policies will add complexity to the Kubewarden controller and the policy server.
+- The expression language used to evaluate the group policies may not be familiar to all users.
+- In the worst case, the group policy evaluation time could be the sum of the evaluation time of all the policies in the group.
+  This is even worse if the group policy contains context-aware policies that introduce additional overhead.
+
+# Alternatives
+
+[alternatives]: #alternatives
+
+# Unresolved questions
+
+[unresolved]: #unresolved-questions
```

Comment:
Do we plan to do some validation when a `ClusterAdmissionPolicyGroup` is created?

Things like:

1. ensure the expression is valid
2. ensure the policies mixed together evaluate compatible types of resources

The 1st issue could be addressed, but there's the problem that Rhai is not available for Go. We could use CEL, given we plan to customize Rhai to look and behave like CEL, but there's no way to strip away CEL functions from it. What I mean is that `"foo".startsWith("f") && policy_1() || policy_2()` is a good CEL expression, but we would panic later on when evaluating that inside of Rhai.

Maybe a solution to this problem would be to create a Wasm module that contains our Rhai validator and use that inside of the controller. That would be quite cool :sunglasses: 


As for the 2nd issue. I fear there's nothing we can do on the short term. We might want to introduce a sidecar in the future that pulls the policies and inspects their metadata.